yksgi/src/lib.rs

```diff
@@ -0,0 +1,344 @@
+        let func = self.module.function(fname.as_ptr());
+        let bb = func.bb(bbidx);
+        let instr = bb.instruction(instridx);
+        let orgaot = get_aot_original(&instr);
+        let ty = instr.get_type();
+        let value = SGValue::new(val, ty);
+        self.frames.last_mut().unwrap().add(instr, value);
+        if let Some(v) = orgaot {
+            self.frames.last_mut().unwrap().add(v, value);
+        }
+    }
+
+    /// Lookup the value of variable `var` in the current frame.
+    unsafe fn lookup(&self, var: &LLVMValue) -> SGValue {
+        if var.is_instruction() {
+            *self.frames.last().unwrap().get(var).unwrap()
+        } else if var.is_constant() {
+            llvmbridge::llvm_const_to_sgvalue(var.0)
+        } else {
+            // GlobalVariable, Function, etc.
+            todo!()
+        }
+    }
+
+    /// Lookup the value of variable `var` in the current frame.
+    unsafe fn setvar(&mut self, var: LLVMValue, val: SGValue) {
+        if var.is_instruction() {
+            self.frames.last_mut().unwrap().add(var, val);
+        } else {
+            // GlobalVariable, Function, etc.
+            todo!()
+        }
+    }
+
+    /// Interpret LLVM IR from the interpreters initialised position. Returns true if the control
+    /// point was reached, or false if the interpreter left the main interpreter loop.
+    pub unsafe fn interpret(&mut self) -> bool {
+        // We start interpretation at the branch instruction that was turned into a guard. We
+        // need to re-interpret this instruction in order to find out which branch we need to
+        // follow.
+        loop {
+            match self.pc.opcode() {
+                LLVMOpcode::LLVMAdd => self.add(),
+                LLVMOpcode::LLVMBitCast => self.bitcast(),
+                LLVMOpcode::LLVMBr => {
+                    self.branch();
+                    continue;
+                }
+                LLVMOpcode::LLVMCall => {
+                    self.call()
+                },
+                LLVMOpcode::LLVMGetElementPtr => self.gep(),
+                LLVMOpcode::LLVMLoad => self.load(),
+                LLVMOpcode::LLVMPHI => self.phi(),
+                LLVMOpcode::LLVMRet => {
+                    if self.ret() {
+                        return true;
+                    }
+                },
+                LLVMOpcode::LLVMStore => self.store(),
+                LLVMOpcode::LLVMSwitch => {
+                    self.switch();
+                    continue;
+                }
+                _ => todo!("{:?}", self.pc.as_str()),
+            }
+            self.pc = LLVMValue(LLVMGetNextInstruction(self.pc.0));
+        }
+    }
+
+    unsafe fn add(&mut self) {
+        let op1 = self.pc.get_operand(0);
+        let op2 = self.pc.get_operand(1);
+        let val1 = self.lookup(&op1);
+        let val2 = self.lookup(&op2);
+        // FIXME: Can overflows happen? What do we do?
```

Comment:
Good question!

It depends if the `add` was qualified with `nuw` or `nsw` (no unsigned overflow, no signed overflow).

https://llvm.org/docs/LangRef.html#add-instruction